### PR TITLE
Clingen-CGC-VICC oncogenicity classification

### LINKED ISSUE #4967: Add support for classifying oncogenicity according to ClinGen-CGC-VICC Guidelines
Currenty, the user can only classify a variant according to the ACMG guidelines. However, for somatic variants it would be good to also have support for the [ClinGen-CGC-VICC Guidelines](https://pubmed.ncbi.nlm.nih.gov/35101336/)

A solution would be to change the classification system for cancer cases (or have the option to classify according to both guidlines?)

### PR DESCRIPTION
**Pull request by molucorner**
This PR adds a functionality or fixes a bug.
Closes #4967

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by CR
- [x] tests executed by CR

### CODE REVIEW

**Comment by northwestwitch** [comment]
Amazing new feature! I can review this one if you haven't started @dnil .

Perhaps we could create a new patch release with the stuff we have merged so far before merging this one?

**Comment by northwestwitch** [comment]
I've started testing the functionality and I've noticed that the general report crashes if you have a cancer SNV variant classified both using ACMG and Clingen-CGC-VICC.

To reproduce, go on demo cancer page and classify it for instance like this:

image


Then if you go on general report page you should get this error:

```
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py39/lib/python3.9/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py39/lib/python3.9/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py39/lib/python3.9/site-packages/flask_cors/extension.py", line 178, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py39/lib/python3.9/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py39/lib/python3.9/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/utils.py", line 87, in decorated_function
    context = func(*args, **kwargs)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/cases/views.py", line 267, in case_report
    data = controllers.case_report_content(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/cases/controllers.py", line 713, in case_report_content
    case_report_variants(store, case_obj, institute_obj, data)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/cases/controllers.py", line 634, in case_report_variants
    _append_evaluated_variant_by_type(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/cases/controllers.py", line 667, in _append_evaluated_variant_by_type
    _get_decorated_var(var_obj=var_obj, institute_obj=institute_obj, case_obj=case_obj)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/cases/controllers.py", line 643, in _get_decorated_var
    return variant_decorator(
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variant/controllers.py", line 329, in variant
    variant_obj["filters"] = get_filters(variant_obj)
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variant/utils.py", line 641, in get_filters
    return [
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variant/utils.py", line 641, in <listcomp>
    return [
  File "/Users/chiararasi/Documents/work/GITs/scout/scout/server/blueprints/variant/utils.py", line 651, in <lambda>
    for f in map(lambda x: x.lower(), variant_obj["filters"])
AttributeError: 'dict' object has no attribute 'lower'
```

**Comment by dnil** [comment]
> Amazing new feature! I can review this one if you haven't started @dnil .
> 
> Perhaps we could create a new patch release with the stuff we have merged so far before merging this one?

Agreed, good stuff @molucorner! I had a Quick Look, and only saw some things that should be simple to fix. I agree with a patch release first; looking at your ClinVar tags up-prio PR https://github.com/Clinical-Genomics/scout/pull/5006. I'm still not convinced it's really a bug, more of an incompletely documented feature, but let's talk about that over there. Maybe we can include it in the patch release  as well.

**Comment by dnil** [comment]
> The error might not be related to the code of this PR specifically, but it's not triggered by classifying a cancer variant using ACMG (tested). Weird!
 
A bit too much of a coincidence it touched the filter tags, isn't it? I didn't test the general report manually with that PR, only what we have in automated tests for it. I'll take a look!

**Review by northwestwitch — CHANGES_REQUESTED** [review]
Good job! I have a few comments, easy stuff to fix!

- My main concern is the report crashing when you have a double classification (see my other comment)

- Would also be nice to address those failed tests due to the sorting of imports!


We should also write tests for this new code, but since this PR is already quite large we can perhaps address this in a new PR. We could basically copy/paste the tests written for the ACMG functionality with very few changes.

**Comment by dnil** [comment]
> > > The error might not be related to the code of this PR specifically, but it's not triggered by classifying a cancer variant using ACMG (tested). Weird!
> > 
> > 
> > A bit too much of a coincidence it touched the filter tags, isn't it? I didn't test the general report manually with that PR, only what we have in automated tests for it. I'll take a look!
> 
> Yes, see #5075; my bad!

Ok, try again now - the general report error from multiple classifications should be suppressed again. Sorry, that endpoint is due for a refactor.

File: scout/server/blueprints/variant/utils.py
Comment (dnil) [reply scout/server/blueprints/variant/utils.py:5]: We also run `black` with a line length of 100. But we have a write enabled linter action to fix that; we haven't fixed the `isort` since it went unsupported by an update to that linter action. Maybe we will switch to `ruff` in a bit https://docs.astral.sh/ruff/formatter/#sorting-imports. 😁

File: scout/server/blueprints/variant/views.py
Comment (molucorner) [reply scout/server/blueprints/variant/views.py:428]: So about this suggestion, it doesn't seem to be used for the other redirects? Should I ignore it?

Comment (northwestwitch) [reply scout/server/blueprints/variant/views.py:428]: Yes, I'd say ignore it, we have this issue in many other endpoints and we might as well fix them all in a dedicated pull request

**Review by northwestwitch — APPROVED** [review]
I think this is ready to be merged now. What do you think @dnil?

Great contribution! 🥇 

Perhaps you could add some tests in another PR?

**Comment by molucorner** [comment]
Sure, I can do the tests! I made a start on it at some point already

**Comment by dnil** [comment]
> I think this is ready to be merged now. What do you think @dnil?
> 
> Great contribution! 🥇
> 
> Perhaps you could add some tests in another PR?

I think it is super! Since you ask, I'm afraid we will get a reviewer 2 situation, but lets see if I find something en passant, and we just do what we always do with new features: patch them as needed! 😁

File: scout/constants/ccv.py
Comment (dnil) [inline scout/constants/ccv.py:50]: I dont think that was intentionally kept? 
```suggestion
```

File: scout/constants/ccv.py
Comment (dnil) [inline scout/constants/ccv.py:218]: 🥇

File: scout/server/blueprints/variant/templates/variant/components.html
Comment (dnil) [inline scout/server/blueprints/variant/templates/variant/components.html:136]: We could make these clickable links, but hey, thats really just a QualityOfLife thing. Great to have the tooltip!

File: scout/utils/ccv.py
Comment (dnil) [inline scout/utils/ccv.py:116]: 🥇

**Review by dnil — APPROVED** [review]
Great work @molucorner! Really well done; I will try to sell it to the local CLGs as well. 😁

**Comment by dnil** [comment]
> I think this is ready to be merged now. What do you think @dnil?
> 
> Great contribution! 🥇
> 
> Perhaps you could add some tests in another PR?

And yes, tests will be needed, but it's quite ok to do it in two steps.

File: scout/constants/ccv.py
Comment (molucorner) [reply scout/constants/ccv.py:50]: Actually, it was kept there on the suggestion of a doctor I've been working with. Personally I don't know enough to have an opinion on it

**Comment by molucorner** [comment]
I think I have all the tests (equivalent to the acmg ones), so does it make more sense to just push them here now?

**Comment by northwestwitch** [comment]
I've deployed this PR on our stage server to give it a final try before merging - the idea was to include it in the upcoming release -  Unfortunately we've noticed a strange behaviour: a variant with the oncogenicity classification that does not appear as such (or at all) on the general case report. For this reason perhaps it's best to give it some more test next week. So we are not including this in the next release but in the one immediately after. Hope it is fine @molucorner

**Comment by molucorner** [comment]
> I've deployed this PR on our stage server to give it a final try before merging - the idea was to include it in the upcoming release - Unfortunately we've noticed a strange behaviour: a variant with the oncogenicity classification that does not appear as such (or at all) on the general case report. For this reason perhaps it's best to give it some try test next week. So we are not including this in the next release but in the one immediately after. Hope it is fine @molucorner

Yes, of course. Let's make it work properly first!

**Comment by northwestwitch** [comment]
I've re-deployed this PR on our stage server now. Let's check why [this classified variant](https://scout-stage.scilifelab.se/cust000/KMP-00064T-20191996305/cancer/db3b01ea8ce1c50e84e1a1f607e36be0?cancer=yes) doesn't show on general report

**Comment by northwestwitch** [comment]
The same case has also this other classified variant: https://scout-stage.scilifelab.se/cust000/KMP-00064T-20191996305/cancer/80b23622ad583a120ea858e9c5977ea7?cancer=yes. Doesn't show on report either .. 🤔

**Comment by northwestwitch** [comment]
I think I found the problem and it's not caused by the code of the PR - another bug, sorry, again! - I will open a new issue describing the problem and the fix. @molucorner feel free to merge this PR when you are ready. Please update the changelog moving the description of this PR upper in the release history. Thanks again for the great contribution! 🙏🏻 🚀

**Comment by molucorner** [comment]
> I think I found the problem and it's not caused by the code of the PR - another bug, sorry, again! - I will open a new issue describing the problem and the fix. @molucorner feel free to merge this PR when you are ready. Please update the changelog moving the description of this PR upper in the release history. Thanks again for the great contribution! 🙏🏻 🚀

Great! Just to be sure I don't mess up anything at the last moment, I should:
      1. Move the description to "unreleased" in the changelog.
      2. Simply click the "Merge pull request" button at the bottom.

As you have probably been able to tell by now, I'm not used to the PR workflow :sweat_smile:. Thanks for your patience, it's been a fun thing to work on!

**Comment by northwestwitch** [comment]
> Great! Just to be sure I don't mess up anything at the last moment, I should:       1. Move the description to "unreleased" in the changelog.       2. Simply click the "Merge pull request" button at the bottom.

Exactly like that! And no worries, if something is messed up we fix it later. We test everything again just before a new release.

> As you have probably been able to tell by now, I'm not used to the PR workflow 😅. Thanks for your patience, it's been a fun thing to work on!

You are doing great! 🥇 👏🏻

**Pull request merged**
